src/tools/fuzzing/fuzzing.cpp

```diff
@@ -275,9 +275,71 @@ void TranslateToFuzzReader::setupMemory() {
 void TranslateToFuzzReader::setupHeapTypes() {
   // Start with any existing heap types in the module, which may exist in any
   // initial content we began with.
-  interestingHeapTypes = ModuleUtils::collectHeapTypes(wasm);
+  auto possibleHeapTypes = ModuleUtils::collectHeapTypes(wasm);
 
   // TODO: use heap type fuzzer to add new types in addition to the previous
+
+  // Filter away uninhabitable heap types, that is, heap types that we cannot
+  // construct, like a type with a non-nullable reference to itself. For
```

Comment:
I don't see where cycles are actually detected here.

Reply:
Oh, I see, it's depending on hitting the recursion limit to detect cycles. That seems wasteful!

Reply:
It's very simple though :smile: look at how little code it is. With a reasonable recursion limit it should be cheap.